### PR TITLE
Machine learning updates

### PR DESCRIPTION
**Pull request by procaconsul**
**Overlapping Block Marker**

This marker is more of a side experiment to verify the possibility of classification and learning improvement given by splitting the source(s) in overlapping blocks (rather than sequential blocks, like we did until now).
The parts of the micro service are tested in Tests.py with the unittest library. 
For the database update, the files written by Giulio for Simple Haskell Marker have been recycled (I needed the same exact things).
With the training files we used for Simple Haskell Marker the classification we get is rather unreliable. The logic of the microservice seems consistent, so the reason for this is probably in the overlapping nature of the blocks. A dynamic creation of the training set (i.e. updated by the user from scratch) is the best solution.

### CODE REVIEW

**Comment by giuliojiang** [comment]
So this new microservice doesn't get called in the full pipeline? I don't see changes to microservices.py

File: backend/overlapping-block-marker/Classifier.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        try: 
+            self.clf.fit(self.X, self.y)
+        except Exception as e:
+            print (str(e))
+            sys.exit(1)
+
+    def predict(self, new_data):
+        prediction = self.clf.predict(new_data)
+        if len(prediction) != len(new_data):
+            raise Exception('[Classifier] Classification failed. Aborting.')
```
Comment (giuliojiang) [inline backend/overlapping-block-marker/Classifier.py:44]: is new_data a string (or equivalently corresponding to the string of source code), and prediction a category? Why should the lengths be the same?

File: backend/overlapping-block-marker/mysqldb.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import SimpleHTTPServer
```
Comment (giuliojiang) [inline backend/overlapping-block-marker/mysqldb.py]: instead of making a copy of this class, use the one in the simple haskell marker directory
```
sys.path.append(path to backend / simple haskell marker)
import mysqldb
```
this way you have the same management of passwords and environment variables

File: infrastructure/infrastructure/src/main/java/alabno/msfeedback/OverBlockMarkerUpdater.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        System.out.println("OverBlockMarkerUpdater:init()");
+
+        // create the directory
+        SubprocessUtils.call("mkdir " + FileUtils.getWorkDir() + "/overlapping-block-marker/training");
+
+        updateTraining();
+    }
+
+    @Override
+    public void update(String source, String type, String annotation) {
```
Comment (giuliojiang) [inline infrastructure/infrastructure/src/main/java/alabno/msfeedback/OverBlockMarkerUpdater.java]: I think this is an outdated interface signature

Comment (procaconsul) [reply infrastructure/infrastructure/src/main/java/alabno/msfeedback/OverBlockMarkerUpdater.java]: Yes, updating

File: backend/overlapping-block-marker/Classifier.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        self.clf = MLPClassifier(solver='adam', alpha=1e-5, 
+                                   hidden_layer_sizes=(5, 2), random_state=1)
+        
+        try: 
+            self.clf.fit(self.X, self.y)
+        except Exception as e:
+            print (str(e))
+            sys.exit(1)
+
+    def predict(self, new_data):
```
Comment (helicopter88) [inline backend/overlapping-block-marker/Classifier.py:40]: I have no idea about what does this method do.

File: backend/overlapping-block-marker/IncrementSerializedClassifier.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import sys
+
+# alabno/simple-haskell-marker
+exec_dir = os.path.abspath(os.path.dirname(os.path.abspath(__file__)))
+
+# get list of items in training directory
+dir_content = os.listdir(exec_dir + os.sep + 'training')
+
+# sort the list
+dir_content.sort()
```
Comment (helicopter88) [inline backend/overlapping-block-marker/IncrementSerializedClassifier.py:61]: why?

File: backend/overlapping-block-marker/Classifier.py (hunk 1 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/Classifier.py:44]: So, new_data is a matrix (i.e. something like [[2.0,35.0, ...], [4.0,9.0,...], ...]) and prediction is an array of categories, each corresponding to every elemento of new_data, in order.
This is why they must have the same length. IF they don't, something has not been classified.

File: backend/overlapping-block-marker/MLUtils.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+# Pad strings
+def pad(line, size):
+    _line = line
+    while len(_line) < size:
+        _line += ' '
+    return line
+    
+# Pad lists of floats
+def pad_float(a_list, size):
```
Comment (helicopter88) [inline backend/overlapping-block-marker/MLUtils.py:41]: Why are you padding it with 32?

File: backend/overlapping-block-marker/Marker.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        
+            to_classify = []
+            for block in blocks:
+                block.format_content()
+                to_classify.append(block.formatted_content)
+            
+            guesses = []
+            # Classify
+            try:
+                guesses = self.classifier.predict(to_classify)
```
Comment (helicopter88) [inline backend/overlapping-block-marker/Marker.py:144]: What does this do?

**Comment by helicopter88** [comment]
Even if we don't wire it, it would be nice to have some sort of benchmark of this compared with MarkMarker or SimpleHaskellMarker(or both)

File: backend/overlapping-block-marker/MLUtils.py (hunk 6 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/MLUtils.py:41]: I have doubts about this being the optimal thing to do. 32.0 is the ASCII value for space, which is what we pad text lines with when they are shorter than needed. I say "needed" because the Classifier takes only input samples of same length.

File: backend/overlapping-block-marker/IncrementSerializedClassifier.py (hunk 5 above)
Comment (giuliojiang) [reply backend/overlapping-block-marker/IncrementSerializedClassifier.py:61]: to find the latest

File: backend/overlapping-block-marker/Classifier.py (hunk 4 above)
Comment (giuliojiang) [reply backend/overlapping-block-marker/Classifier.py:40]: takes in a piece of source code, and returns a category corresponding to an error or annotation or ok (my understanding)

File: backend/overlapping-block-marker/IncrementSerializedClassifier.py (hunk 5 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/IncrementSerializedClassifier.py:61]: This is G's work tbh, I recycled it. I believe it's because training files have incremental names (numerically), so sorting them gives you back the files from the oldest to the youngest?
@giuliojiang

File: backend/overlapping-block-marker/MLUtils.py (hunk 6 above)
Comment (giuliojiang) [reply backend/overlapping-block-marker/MLUtils.py:41]: ascii 32 is whitespace

Comment (helicopter88) [reply backend/overlapping-block-marker/MLUtils.py:41]: I think you want something else, pheraps similar to \0 in concept

File: backend/overlapping-block-marker/Classifier.py (hunk 1 above)
Comment (giuliojiang) [reply backend/overlapping-block-marker/Classifier.py:44]: ok I see that new_data is all the blocks, so each block has to correspond to a prediction

File: backend/overlapping-block-marker/Classifier.py (hunk 4 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/Classifier.py:40]: It's equivalent to "classify" for standford classifier

File: backend/overlapping-block-marker/Marker.py (hunk 7 above)
Comment (giuliojiang) [reply backend/overlapping-block-marker/Marker.py:144]: to_classify should be the array of the input source split into blocks
guesses is the array containing all the predicted annotations and errors

**Comment by procaconsul** [comment]
@helicopter88 @giuliojiang Yes, benchmarking should be needed. I'd like to wire it in fact to see if it is actually suitable for everything or not, then we can more easily benchmark too.

**Comment by giuliojiang** [comment]
I think it's hard to get benchmarking data related to 'accuracy' and stuff because annotating is a very human task. Probably some examples showing input source code, generated annotation and relevant section of training data could be useful

**Comment by procaconsul** [comment]
@giuliojiang yes, and for this data to be relevant once generated by the marker we need to train it as per description of the PR

**Comment by helicopter88** [comment]
Some very basic form of benchmarking (that would allow it to be compared with MarkMarker) is just taking the grade.
I have uploaded on slack some benchmark data coming from MarkMarker, Linter and ComplexityAnalyser, if you use the same files with this one you may be able to produce something that can be used to evaluate.

File: backend/overlapping-block-marker/Classifier.py (hunk 1 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/Classifier.py:44]: Correct

File: backend/overlapping-block-marker/MLUtils.py (hunk 6 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/MLUtils.py:41]: Probably, even though in terms of the ML it should not change that much. Or should it maybe?

File: backend/overlapping-block-marker/Marker.py (hunk 7 above)
Comment (procaconsul) [reply backend/overlapping-block-marker/Marker.py:144]: [...] all the predicted *categories associated to the blocks. I will describe the way this all works in the description, maybe that's better.

**Review by giuliojiang — APPROVED** [review]

**Comment by helicopter88** [comment]
What kind of feedback does this service give?

**Comment by procaconsul** [comment]
@helicopter88 What do you mean for feedback? The output? Just like the others, it returns annotations.

**Comment by helicopter88** [comment]
I mean, does it assess style? does it assess nothing like MarkMarker? does it give the same output as SimpleHaskellMarker?

**Comment by procaconsul** [comment]
@helicopter88 Its output is just like the one produced by the Simple Haskell Marker, correct.
